### ISSUE TITLE
Try not to allocate ports that are in use

### ISSUE THREAD

**Issue by seandst** [seandst/webdriver-wharf#5]
https://github.com/seandst/webdriver-wharf/blob/master/webdriver_wharf/interactions.py#L53

At the moment, wharf tries not to allocate ports that it's already using, but has no regard for other services that may be listening on the box. `_next_available_port` should make an attempt to check those ports before binding (which will create a nice race condition between checking the port and actually binding it, but is probably still an improvement...?)).

Additionally, it would probably good to decouple the ports (e.g. container 17 doesn't always get ports x917), and check (un)availability for each individual  service.

**Comment by jchristi** [comment]
It would be good to reduce the overall number of ports in use by re-using the same port for the same service on every container and use sub-directories to separate them. That probably can't be done for some services like VNC (or could it?), but would certainly help with port usage (and firewall management).

**Comment by jchristi** [comment]
I've been thinking about this for awhile and here's what I've come up with. Docker has the ability to publish and bind all exposed ports to ephemeral port on the host:

  > you can supply `-P` or `--publish-all=true|false` to `docker run` which is a blanket operation that identifies every port with an `EXPOSE` line in the image’s `Dockerfile` or `--expose <port>` commandline flag and maps it to a host port somewhere within an ephemeral port range. The `docker port` command then needs to be used to inspect created mapping.

  via [Docker Docs: Bind container ports to the host](https://docs.docker.com/engine/userguide/networking/default_network/binding/)

This pretty much solves all of our port-related problems. No worries about race conditions or ports not being available since they'll be allocated by the OS via docker. And this means no lower-range ports will be taken up and no broad firewall rules need to be made to the host system. It also eliminates my concern and desire for re-using the same port for multiple containers.

I may spend some time hacking at this while on vacation. Maybe.

**Comment by seandst** [comment]
> no broad firewall rules need to be made to the host system

Except allowing ingress on the entire ephemeral port range for new connections, right? That's the only reason wharf hasn't done this from the start, iirc.

**Comment by jchristi** [comment]
I think it's better than the current solution of using lower range ports. With the current solution, you have to open up a large amount of lower range ports that wharf never uses, because of the offsets. In either case you could change it to use only ports within a specific range and no offsets. But wharf containers are ephemeral so it should follow that they also use ephemeral ports. The range of available ports is much larger in the ephemeral range, but that can be reduced via configuration as desired.

**Comment by RonnyPfannschmidt** [comment]
i believe #13 fixed this as well by extracting the ports from docker instead of imposing the ports to docker

**Comment by RonnyPfannschmidt** [comment]
yup - its being facilitated in https://github.com/RedHatQE/webdriver-wharf/pull/13/files#diff-d4cd6936ae970fe41469e6aa803a742bR61
